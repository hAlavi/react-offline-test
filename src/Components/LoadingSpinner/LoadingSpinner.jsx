import styled from 'styled-components';
import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';

export const PageOverlay = styled.div`
  position: fixed;
  left: 0;
  top: 0;
  z-index: 99999;
  display: flex;
  align-items: center;
  width: 100vw;
  height: 100vh;
  background-color: #999999;
`;

export const LoadingSpinnerIcon = styled.div`
  margin: auto;
  font-size: 32px;
  @keyframes spin {
    0% {
      -webkit-transform: rotate(0deg);
      -ms-transform: rotate(0deg);
      transform: rotate(0deg);
    }
    100% {
      -webkit-transform: rotate(360deg);
      -ms-transform: rotate(360deg);
      transform: rotate(360deg);
    }
  }

  animation: spin 2s infinite linear;
`;

const LoadingSpinner = () => {
  return (
    <PageOverlay>
      <LoadingSpinnerIcon data-testid="loading-spinner">
        <FontAwesomeIcon icon={faSpinner} />
      </LoadingSpinnerIcon>
    </PageOverlay>
  );
};

export default LoadingSpinner;
