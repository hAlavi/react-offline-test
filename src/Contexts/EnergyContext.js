import React from 'react';
import PropTypes from 'prop-types';
import axios from '../Helpers/axios';
import config from '../Helpers/config';
import defaultState from './example_api_response.json';

const EnergyContext = React.createContext(defaultState);

class EnergyProvider extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = { ...props.energyContext };
  }

  retrieveContext = () => {
    this.setState({
      loading: true,
    });

    axios
      .get(config.energy_service_url)
      .then(response => {
        this.setState({
          loading: false,
          energy: {
            ...response.data.body,
          },
        });
      })
      .catch(error => {
        const errorDetails = {};
        if (error.response) {
          errorDetails.errorCode = error.response.status;
          errorDetails.errorMessage = error.response.data.errorMessage;
        } else {
          errorDetails.errorCode = -1;
          errorDetails.errorMessage = `Request to retrieve energy details failed due to ${error.message}`;
        }

        this.setState({
          loading: false,
          energy: {
            ...errorDetails,
          },
        });
      });
  };

  render() {
    const { children } = this.props;
    return (
      <EnergyContext.Provider
        value={{
          ...this.state,
          retrieveContext: this.retrieveContext,
        }}
      >
        {children}
      </EnergyContext.Provider>
    );
  }
}

EnergyProvider.propTypes = {
  children: PropTypes.node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  energyContext: PropTypes.object,
};

EnergyProvider.defaultProps = {
  energyContext: defaultState,
};

const EnergyConsumer = EnergyContext.Consumer;

export { EnergyProvider, EnergyConsumer };
