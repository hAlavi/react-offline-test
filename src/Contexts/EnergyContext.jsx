import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import axios from '../Helpers/axios';
import config from '../Helpers/config';
import log from '../Helpers/logger';

const ErrorMessage = styled.p`
  color: red;
  font-size: 24px;
`;

const NS_CONTEXT = 'KiwiPower:EnergyContext';

// Uncomment this block to use sample response as a defaultState
// import defaultResponse from './example_api_response.json';
// const defaultState = {
//   loading: false,
//   energy: defaultResponse,
//   retrieveEnergy: () => {},
// };

// Default State of the context
const defaultState = {
  loading: false,
  failed: false,
  energy: {
    data: {
      generationmix: [
        { fuel: 'testFuel1', perc: 10 },
        { fuel: 'testFuel2', perc: 20 },
      ],
    },
  },
  retrieveEnergy: () => {},
};

const EnergyContext = React.createContext(defaultState);

/**
 * Class EnergyProvider to pass the data to the consumer component
 * @see React.createContext
 */
class EnergyProvider extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = { ...props.energy };
  }

  retrieveEnergy = () => {
    this.setState({
      loading: true,
    });

    axios
      .get(config.energy_service_url)
      .then(response => {
        log(NS_CONTEXT, response);
        this.setState({
          loading: false,
          failed: false,
          energy: {
            ...response.data,
          },
        });
      })
      .catch(error => {
        const errorDetails = {};
        //
        if (error.response) {
          errorDetails.errorCode = error.response.status;
          errorDetails.errorMessage = error.response.data.errorMessage;
        } else {
          errorDetails.errorCode = -1;
          errorDetails.errorMessage = `Request to retrieve energy details failed due to ${error.message}`;
        }
        log(NS_CONTEXT, errorDetails.errorMessage);

        this.setState({
          loading: false,
          failed: true,
          energy: {
            ...errorDetails,
          },
        });
      });
  };

  render() {
    const { children } = this.props;
    const { failed } = this.state;

    if (failed) {
      const {
        energy: { errorMessage },
      } = this.state;
      return (
        <ErrorMessage data-testid="error-message">{errorMessage}</ErrorMessage>
      );
    }
    return (
      <EnergyContext.Provider
        value={{
          ...this.state,
          retrieveEnergy: this.retrieveEnergy,
        }}
      >
        {children}
      </EnergyContext.Provider>
    );
  }
}

EnergyProvider.propTypes = {
  children: PropTypes.node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  energy: PropTypes.object,
};

EnergyProvider.defaultProps = {
  energy: defaultState,
};

const EnergyConsumer = EnergyContext.Consumer;

export { EnergyProvider, EnergyConsumer };
