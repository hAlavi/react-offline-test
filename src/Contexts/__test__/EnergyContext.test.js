/* eslint-disable react/prop-types */
import React, { useEffect } from 'react';
import {
  render,
  cleanup,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { EnergyProvider, EnergyConsumer } from '../EnergyContext';
import LoadingSpinner from '../../Components/LoadingSpinner/LoadingSpinner';
import axios from '../../Helpers/axios';
import 'jest-styled-components';

jest.mock('../../Helpers/axios');

// Simple Consumer to test the provider/consumer
const SimpleConsumer = ({ energy }) => <>{JSON.stringify(energy)}</>;

// Component to trigger axios mock to get the context
const ConsumerWithApi = ({ energy }) => {
  const { retrieveEnergy } = energy;
  useEffect(() => {
    retrieveEnergy();
  }, [retrieveEnergy]);

  if (energy.loading) {
    return <LoadingSpinner />;
  }
  return (
    <div data-testid="context-data">
      <div>{energy.energy.data.generationmix[0].fuel}</div>
      <div>{energy.energy.data.generationmix[1].fuel}</div>
    </div>
  );
};

const mockData = {
  data: {
    generationmix: [
      { fuel: 'testFuelMock1', perc: 1 },
      { fuel: 'testFuelMock2', perc: 2 },
    ],
  },
};

afterEach(cleanup);

describe('Unit tests for the EnergyContext', () => {
  describe('Visual Snapshot testing', () => {
    it('Renders without crashing', () => {
      const { container } = render(
        <EnergyProvider>
          <EnergyConsumer>
            {energy => <SimpleConsumer energy={energy} />}
          </EnergyConsumer>
        </EnergyProvider>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Functional unit tests', () => {
    it('it applies context to the consumer - Happy path', async () => {
      axios.get.mockImplementation(() => {
        return Promise.resolve({
          status: 200,
          data: mockData,
        });
      });
      const { getByTestId, getByText } = render(
        <EnergyProvider>
          <EnergyConsumer>
            {energy => <ConsumerWithApi energy={energy} />}
          </EnergyConsumer>
        </EnergyProvider>,
      );

      await waitForElementToBeRemoved(() => getByTestId('loading-spinner'));

      expect(getByTestId('context-data')).toBeTruthy();
      expect(getByText('testFuelMock1')).toBeTruthy();
      expect(getByText('testFuelMock2')).toBeTruthy();
    });
    it('it applies context to the consumer, should return Error with Response - Unhappy path', async () => {
      axios.get.mockImplementation(() => {
        // eslint-disable-next-line prefer-promise-reject-errors
        return Promise.reject({
          response: {
            status: 400,
            data: {
              errorMessage: 'ERROR',
            },
          },
        });
      });
      const { getByTestId } = render(
        <EnergyProvider>
          <EnergyConsumer>
            {energy => <ConsumerWithApi energy={energy} />}
          </EnergyConsumer>
        </EnergyProvider>,
      );

      await waitForElementToBeRemoved(() => getByTestId('loading-spinner'));

      expect(getByTestId('error-message')).toBeTruthy();
    });

    it('it applies context to the consumer, should return Error without Response - Unhappy path', async () => {
      axios.get.mockImplementation(() => {
        // eslint-disable-next-line prefer-promise-reject-errors
        return Promise.reject({
          message: 'ERROR',
        });
      });
      const { getByTestId } = render(
        <EnergyProvider>
          <EnergyConsumer>
            {energy => <ConsumerWithApi energy={energy} />}
          </EnergyConsumer>
        </EnergyProvider>,
      );

      await waitForElementToBeRemoved(() => getByTestId('loading-spinner'));

      expect(getByTestId('error-message')).toBeTruthy();
    });
  });
});
