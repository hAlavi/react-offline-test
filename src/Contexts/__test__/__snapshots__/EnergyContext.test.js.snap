// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for the EnergyContext Visual Snapshot testing Renders without crashing 1`] = `{"loading":false,"failed":false,"energy":{"data":{"generationmix":[{"fuel":"testFuel1","perc":10},{"fuel":"testFuel2","perc":20}]}}}`;
